File: 3779.minimum-number-of-operations-to-have-distinct-elements.go
#
# @lc app=leetcode id=3779 lang=golang
#
# [3779] Minimum Number of Operations to Have Distinct Elements
#

# @lc code=start
func minOperations(nums []int) int {
    pos := make(map[int][]int)
    for i, num := range nums {
        pos[num] = append(pos[num], i)
    }
    spoil := -1
    for _, p := range pos {
        if len(p) >= 2 {
            if spoil < p[len(p)-2] {
                spoil = p[len(p)-2]
            }
        }
    }
    i := 0
    ops := 0
    n := len(nums)
    for i < n && i <= spoil {
        ops++
        rem := n - i
        if rem < 3 {
            i += rem
        } else {
            i += 3
        }
    }
    return ops
}
# @lc code=end

File: 3779.minimum-number-of-operations-to-have-distinct-elements.java
#
# @lc app=leetcode id=3779 lang=java
#
# [3779] Minimum Number of Operations to Have Distinct Elements
#

# @lc code=start
import java.util.*;

class Solution {
    public int minOperations(int[] nums) {
        int n = nums.length;
        Map<Integer, List<Integer>> posMap = new HashMap<>();
        for (int i = 0; i < n; i++) {
            posMap.computeIfAbsent(nums[i], k -> new ArrayList<>()).add(i);
        }
        int maxStart = 0;
        for (List<Integer> positions : posMap.values()) {
            if (positions.size() >= 2) {
                maxStart = Math.max(maxStart, positions.get(positions.size() - 2) + 1);
            }
        }
        return (maxStart + 2) / 3;
    }
}
# @lc code=end